### PR TITLE
More refactoring for ScopeManager

### PR DESCRIPTION
**Pull request by tylerbenson**
The main goal here is to enable better modularity in how scopes are managed. A ScopeInterceptor can be added to enable wrapping of the scope to support various behaviors.  This process also allowed splitting up existing logic to more cohesive parts.  It might be easier to follow to view individual commits.

### CODE REVIEW

File: dd-trace-core/src/main/java/datadog/trace/core/scopemanager/DelegateScopeInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
+package datadog.trace.core.scopemanager;
+
+import datadog.trace.bootstrap.instrumentation.api.AgentScope;
+import datadog.trace.bootstrap.instrumentation.api.AgentSpan;
+import datadog.trace.context.TraceScope;
+
+abstract class DelegateScopeInterceptor implements ScopeInterceptor {
+  protected final ScopeInterceptor delegate;
```
Comment (mar-kolya) [inline dd-trace-core/src/main/java/datadog/trace/core/scopemanager/DelegateScopeInterceptor.java:8]: Is this essentially for chaining and not really for 'delegate' in a 'wrap and change behaviour'? It would be nice to have some comments into intent of this design.

**Comment by mar-kolya** [comment]
I did look over 'ScopeEvent` part - looks reasonable, but this code could really benefit from more comments explaining design intent - especially for 'base' classes and interfaces.

File: dd-trace-core/src/test/groovy/datadog/trace/core/scopemanager/ScopeManagerTest.groovy [hunk 2]
```diff
@@ -138,11 +138,11 @@ class ScopeManagerTest extends DDSpecification {
     for (int i = 0; i <= depth; i++) {
       def span = tracer.buildSpan("test").start()
       scope = tracer.activateSpan(span)
-      assert scope instanceof ContinuableScope
+      assert scope instanceof DDScope
     }
 
     then: "last scope is still valid"
-    (scope as ContinuableScope).depth() == depth
+    (scope as DDScope).depth() == depth
```
Comment (randomanderson) [inline dd-trace-core/src/test/groovy/datadog/trace/core/scopemanager/ScopeManagerTest.groovy]: Isn't this assuming a continuable scope? `depth()` isn't on `DDScope` anymore.

File: dd-trace-core/src/test/groovy/datadog/trace/core/scopemanager/ScopeManagerTest.groovy [hunk 3]
```diff
@@ -157,16 +157,16 @@ class ScopeManagerTest extends DDSpecification {
     scope instanceof NoopAgentScope
 
     and: "scope stack not effected."
-    (scopeManager.active() as ContinuableScope).depth() == depth
+    (scopeManager.active() as DDScope).depth() == depth
```
Comment (randomanderson) [inline dd-trace-core/src/test/groovy/datadog/trace/core/scopemanager/ScopeManagerTest.groovy]: See previous comment about `depth()`

File: dd-trace-core/src/main/java/datadog/trace/core/serialization/MsgpackFormatWriter.java [hunk 4]
```diff
@@ -86,16 +93,4 @@ public void writeBigInteger(byte[] key, BigInteger value, MessagePacker destinat
...
-    if (value.length() < 256) {
-      byte[] interned = StringTables.getBytesUTF8(value);
-      if (null != interned) {
-        destination.packRawStringHeader(interned.length);
-        destination.addPayload(interned);
-        return;
-      }
-    }
-    destination.packString(value);
-  }
```
Comment (randomanderson) [inline dd-trace-core/src/main/java/datadog/trace/core/serialization/MsgpackFormatWriter.java:100]: This whole file seems unrelated to the PR

Comment (randomanderson) [reply dd-trace-core/src/main/java/datadog/trace/core/serialization/MsgpackFormatWriter.java:100]: nvm.  I see the commit comment
